### PR TITLE
add basic runtime

### PR DESCRIPTION
**Pull request by BusyJay**
This pr adds basic runtime for failpoints. There still needs a global registry and macros definition to make it useful.

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @brayniac

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+
+fn partition(s: &str, pattern: char) -> (&str, Option<&str>) {
+    let mut splits = s.splitn(2, pattern);
+    (splits.next().unwrap(), splits.next())
+}
+
+impl FromStr for Action {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Action, String> {
```
Comment (siddontang) [inline src/lib.rs]: give you give an arg string format example?

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+            Task::Sleep(t) => thread::sleep(Duration::from_millis(t)),
+            Task::Panic(msg) => match msg {
+                Some(ref msg) => panic!("{}", msg),
+                None => panic!("failpoint {} panic", self.name),
+            },
+            Task::Print(msg) => match msg {
+                Some(ref msg) => println!("{}", msg),
+                None => println!("failpoint {} executed.", self.name),
+            },
+            Task::Pause => unreachable!(),
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/lib.rs]: I don't get it. Why define `Pause`?

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+            Task::Yield => thread::yield_now(),
+            Task::Delay(t) => {
+                let timer = Instant::now();
+                let timeout = Duration::from_millis(t);
+                while timer.elapsed() < timeout {}
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/lib.rs]: That's by design.

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+
+impl FromStr for Action {
+    type Err = String;
+
+    /// Parse an action.
+    ///
+    /// `s` should be in the format `[p%][cnt*]task[(args)]`, `p%` is the frequency,
+    /// `cnt` is the max times the action can be triggered.
+    fn from_str(s: &str) -> Result<Action, String> {
+        let mut remain = s.trim();
```
Comment (siddontang) [inline src/lib.rs]: can we use a regexp to handle this? 

I guess may be easier here.

Comment (BusyJay) [reply src/lib.rs]: I don't want to introduce a heavy and complicated dependency for this simple functionality.

**Comment by BusyJay** [comment]
ping

File: src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    }
+}
+
+// A better name?
+pub type ReturnValue = Option<String>;
+
+pub struct FailPoint {
+    name: &'static str,
+    pause: Mutex<bool>,
+    pause_notifier: Condvar,
```
Comment (siddontang) [inline src/lib.rs]: may be easier to use thread pack and unpack directly.

Comment (BusyJay) [reply src/lib.rs]: Then I need to use a `mutex + vector` to save the thread handles. I think it's more simple to use `Mutex<bool> + Condvar`.

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+impl Action {
+    fn new(task: Task, freq: f32, max_cnt: Option<usize>) -> Action {
+        Action {
+            task: task,
+            freq: freq,
+            count: max_cnt.map(AtomicUsize::new),
+        }
+    }
+
+    fn get_task(&self) -> Option<Task> {
```
Comment (overvenus) [inline src/lib.rs]: How about `maybe_get_task`? Sin it may not return a task even if there is one.

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+        let point = FailPoint::new("test_fail_point_panic");
+        point.set_actions(vec![Action::new(Task::Panic(None), 1.0, None)]);
+        point.eval();
+    }
+
+    #[test]
+    fn test_print() {
+        let point = FailPoint::new("test_fail_point_print");
+        point.set_actions(vec![Action::new(Task::Print(None), 1.0, None)]);
+        assert!(point.eval().is_none());
```
Comment (overvenus) [inline src/lib.rs]: Please check stdout.

**Comment by siddontang** [comment]
LGTM

File: src/lib.rs (hunk 6 above)
Comment (BusyJay) [reply src/lib.rs]: I think `get` is ok. For example, in stdlib, there are many `get` methods return `Option`, but no method named `maybe_get`.

File: src/lib.rs (hunk 7 above)
Comment (BusyJay) [reply src/lib.rs]: I may change it to log latter, so let's check it then.

**Comment by overvenus** [comment]
LGTM

**Pull request merged**